File: ios/RecipeClipper/UI/Plan/PlanComponents.swift
import SwiftUI

/// Day labels (#49). A day is an epoch day, so it is formatted at midnight UTC with a UTC
/// calendar: the local zone could put that instant on the day before.
enum PlanDayFormat {
    private static func format(_ day: Int64, _ template: String) -> String {
        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.setLocalizedDateFormatFromTemplate(template)
        return formatter.string(from: PlanDays.utcDate(day))
    }

    /// "Monday 23", in the locale's order.
    static func title(_ day: Int64) -> String { format(day, "EEEEd") }

    /// "Mon"
    static func shortWeekday(_ day: Int64) -> String { format(day, "EEE") }

    /// "23"
    static func dayOfMonth(_ day: Int64) -> String { format(day, "d") }

    /// "Sep 21 – 27", or across months "Sep 28 – Oct 4".
    static func weekRange(_ start: Int64) -> String {
        let end = start + 6
        let sameMonth = format(start, "M") == format(end, "M")
        return format(start, "MMMd") + " – " + (sameMonth ? format(end, "d") : format(end, "MMMd"))
    }
}

/// The strip of days in a plan sheet: this week and next, a short weekday over its date. The
/// chosen day is ringed in paprika; today's date is paprika text.
struct DayStrip: View {
    let days: [Int64]
    let today: Int64
    let selected: Int64
    let onSelect: (Int64) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 6) {
                ForEach(days, id: \.self) { day in
                    let isSelected = day == selected
                    Button { onSelect(day) } label: {
                        VStack(spacing: 2) {
                            Text(PlanDayFormat.shortWeekday(day))
                                .textStyle(Typography.labelMedium)
                                .foregroundStyle(Palette.muted)
                            Text(PlanDayFormat.dayOfMonth(day))
                                .textStyle(Typography.titleMedium)
                                .foregroundStyle(day == today ? Palette.accentText : Palette.onBackground)
                        }
                        .frame(minWidth: 48)
                        .padding(.vertical, 8)
                        .padding(.horizontal, 6)
                        .background(
                            RoundedRectangle(cornerRadius: 12)
                                .strokeBorder(isSelected ? Palette.accentText : Palette.hairline, lineWidth: isSelected ? 2 : 1)
                        )
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(PlanDayFormat.title(day))
                    .accessibilityAddTraits(isSelected ? [.isSelected] : [])
                    .accessibilityIdentifier("planDay-\(day)")
                }
            }
        }
    }
}

/// The meal types as a row of outlined choices, the chosen one ringed in paprika.
struct MealTypeChoices: View {
    let types: [MealType]
    let selected: Int64?
    let onSelect: (Int64) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(types) { type in
                    let isSelected = type.id == selected
                    Button { onSelect(type.id) } label: {
                        Text(type.name)
                            .textStyle(Typography.labelLarge)
                            .foregroundStyle(isSelected ? Palette.accentText : Palette.onBackground)
                            .padding(.horizontal, 14)
                            .padding(.vertical, 8)
                            .background(
                                Capsule().strokeBorder(isSelected ? Palette.accentText : Palette.hairline, lineWidth: isSelected ? 2 : 1)
                            )
                            .contentShape(Capsule())
                    }
                    .buttonStyle(.plain)
                    .accessibilityAddTraits(isSelected ? [.isSelected] : [])
                }
            }
        }
    }
}

/// "Serves − 4 +", as on the recipe screen.
struct ServingsPicker: View {
    let servings: Int
    let onChange: (Int) -> Void

    var body: some View {
        HStack(spacing: 6) {
            Text(Strings.serves).textStyle(Typography.bodyLarge).foregroundStyle(Palette.onBackground)
            stepButton("−", enabled: servings > 1, label: Strings.decreaseServings) { onChange(servings - 1) }
            Text("\(servings)")
                .textStyle(Typography.titleMedium)
                .foregroundStyle(Palette.onBackground)
                .frame(minWidth: 32)
            stepButton("+", enabled: servings < Servings.max, label: Strings.increaseServings) { onChange(servings + 1) }
        }
    }

    private func stepButton(_ glyph: String, enabled: Bool, label: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(glyph)
                .textStyle(Typography.titleLarge)
                .frame(width: 40, height: 40)
                .background(Circle().fill(Palette.surfaceContainer))
        }
        .buttonStyle(.plain)
        .foregroundStyle(Palette.onBackground)
        .disabled(!enabled)
        .opacity(enabled ? 1 : 0.4)
        .accessibilityLabel(label)
    }
}

/// The body shared by "Add to plan" and the Week's "Move": days, meal type, maybe servings,
/// then one button.
struct PlanSheetContent: View {
    let title: String
    let days: [Int64]
    let today: Int64
    let selectedDay: Int64
    let onDaySelected: (Int64) -> Void
    let mealTypes: [MealType]
    let selectedMealTypeId: Int64?
    let onMealTypeSelected: (Int64) -> Void
    var servings: Int? = nil
    var onServingsChange: (Int) -> Void = { _ in }
    let confirmLabel: String
    let confirmEnabled: Bool
    let onConfirm: () -> Void

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                SectionHeading(title).padding(.bottom, 12)
                DayStrip(days: days, today: today, selected: selectedDay, onSelect: onDaySelected)
                Text(Strings.labelMeal)
                    .textStyle(Typography.labelLarge)
                    .foregroundStyle(Palette.muted)
                    .padding(.top, 20)
                    .padding(.bottom, 8)
                MealTypeChoices(types: mealTypes, selected: selectedMealTypeId, onSelect: onMealTypeSelected)
                if let servings {
                    ServingsPicker(servings: servings, onChange: onServingsChange).padding(.top, 16)
                }
                Button(confirmLabel, action: onConfirm)
                    .buttonStyle(PrimaryButtonStyle())
                    .disabled(!confirmEnabled)
                    .padding(.top, 24)
                    .accessibilityIdentifier("plan.confirm")
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 24)
            .readableColumn()
        }
        .presentationBackground(Palette.background)
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/plan/PlanComponents.kt
package com.example.recipeclipper.ui.plan

import android.icu.text.SimpleDateFormat
import android.icu.util.TimeZone
import android.text.format.DateFormat
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.FilledTonalIconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.example.recipeclipper.R
import com.example.recipeclipper.data.model.MealType
import com.example.recipeclipper.data.model.PlanDays
import com.example.recipeclipper.data.model.Servings
import java.util.Date
import java.util.Locale

// Day labels. A day is an epoch day, so it is formatted at midnight UTC with a UTC calendar:
// the local zone could put that instant on the day before. ICU's SimpleDateFormat, the pair of
// the ICU skeleton pattern: it may hold ICU-only letters ("ccc" for a standalone weekday), which
// Android's java.text copy also takes but the JVM's (so Robolectric's, #91) doesn't.

private fun format(day: Long, skeleton: String): String {
    val locale = Locale.getDefault()
    val formatter = SimpleDateFormat(DateFormat.getBestDateTimePattern(locale, skeleton), locale)
    formatter.timeZone = TimeZone.getTimeZone("UTC")
    return formatter.format(Date(PlanDays.utcMillis(day)))
}

/** "Monday 23", in the locale's order. */
internal fun dayTitle(day: Long): String = format(day, "EEEEd")

/** "Mon" */
internal fun shortWeekday(day: Long): String = format(day, "EEE")

/** "Sep 23", in the locale's order. */
internal fun shortDate(day: Long): String = format(day, "MMMd")

/** "23" */
internal fun dayOfMonth(day: Long): String = format(day, "d")

/** "Sep 21 – 27", or across months "Sep 28 – Oct 4". */
internal fun weekRange(start: Long): String {
    val end = start + 6
    val sameMonth = format(start, "M") == format(end, "M")
    return format(start, "MMMd") + " – " + (if (sameMonth) format(end, "d") else format(end, "MMMd"))
}

/**
 * The strip of days in a plan sheet: this week and next, each a short weekday over its date.
 * The chosen day is ringed in paprika, today's date is paprika text. An exclusive choice, so
 * each cell is `selectable` with a radio-button role.
 */
@Composable
internal fun DayStrip(days: List<Long>, today: Long, selected: Long, onSelect: (Long) -> Unit) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        modifier = Modifier.horizontalScroll(rememberScrollState())
    ) {
        days.forEach { day ->
            val isSelected = day == selected
            val shape = RoundedCornerShape(12.dp)
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier
                    .widthIn(min = 48.dp)
                    .border(
                        BorderStroke(
                            if (isSelected) 2.dp else 1.dp,
                            if (isSelected) MaterialTheme.colorScheme.tertiary else MaterialTheme.colorScheme.outlineVariant
                        ),
                        shape
                    )
                    .selectable(selected = isSelected, role = Role.RadioButton, onClick = { onSelect(day) })
                    .padding(vertical = 8.dp, horizontal = 6.dp)
                    .testTag("planDay-$day")
            ) {
                Text(
                    shortWeekday(day),
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    dayOfMonth(day),
                    style = MaterialTheme.typography.titleMedium,
                    color = if (day == today) MaterialTheme.colorScheme.tertiary else MaterialTheme.colorScheme.onSurface
                )
            }
        }
    }
}

/** The meal types as a row of outlined choices, the chosen one ringed in paprika. */
@Composable
internal fun MealTypeChoices(types: List<MealType>, selected: Long?, onSelect: (Long) -> Unit) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier.horizontalScroll(rememberScrollState())
    ) {
        types.forEach { type ->
            val isSelected = type.id == selected
            Surface(
                shape = RoundedCornerShape(50),
                color = Color.Transparent,
                border = BorderStroke(
                    if (isSelected) 2.dp else 1.dp,
                    if (isSelected) MaterialTheme.colorScheme.tertiary else MaterialTheme.colorScheme.outlineVariant
                ),
                modifier = Modifier.selectable(selected = isSelected, role = Role.RadioButton, onClick = { onSelect(type.id) })
            ) {
                Text(
                    type.name,
                    style = MaterialTheme.typography.labelLarge,
                    color = if (isSelected) MaterialTheme.colorScheme.tertiary else MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.padding(horizontal = 14.dp, vertical = 8.dp)
                )
            }
        }
    }
}

/** "Serves − 4 +", as on the recipe screen. */
@Composable
internal fun ServingsPicker(servings: Int, onChange: (Int) -> Unit) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Text(stringResource(R.string.label_serves), style = MaterialTheme.typography.bodyLarge)
        Spacer(Modifier.width(6.dp))
        val decrease = stringResource(R.string.cd_decrease_servings)
        FilledTonalIconButton(
            onClick = { onChange(servings - 1) },
            enabled = servings > 1,
            modifier = Modifier.semantics { contentDescription = decrease }
        ) { Text("−", style = MaterialTheme.typography.titleLarge) }
        Text(
            "$servings",
            style = MaterialTheme.typography.titleMedium,
            textAlign = TextAlign.Center,
            modifier = Modifier.widthIn(min = 32.dp)
        )
        val increase = stringResource(R.string.cd_increase_servings)
        FilledTonalIconButton(
            onClick = { onChange(servings + 1) },
            enabled = servings < Servings.MAX,
            modifier = Modifier.semantics { contentDescription = increase }
        ) { Text("+", style = MaterialTheme.typography.titleLarge) }
    }
}

/** Remembers [dayTitle] per day, so a list of days isn't reformatting on every frame. */
@Composable
internal fun rememberDayTitle(day: Long): String = remember(day) { dayTitle(day) }
